include/pybind11/cast.h

```diff
@@ -1055,6 +1059,16 @@ T cast(const handle &handle) {
     return T(reinterpret_borrow<object>(handle));
 }
 
+// Note that `cast<PyObject *>(obj)` increments the reference count of `obj`.
+// This is necessary for the case that `obj` is a temporary.
+// It is the responsibility of the caller to ensure that the reference count
+// is decremented.
+template <typename T,
+          detail::enable_if_t<detail::is_same_ignoring_cvref<T, PyObject *>::value, int> = 0>
+T cast(const handle &handle) {
+    return handle.inc_ref().ptr();
+}
```

Comment:
Is the any case where you want to steal the ref? Ie. shouldn't you also add this?
```suggestion
}
template <typename T,
          detail::enable_if_t<detail::is_same_ignoring_cvref<T, PyObject *>::value, int> = 0>
T cast(object &&obj) {
    return obj.release().ptr();
}
```

Reply:
> Is the any case where you want to steal the ref?

Not sure. I think the best way to answer this: can we come up with a meaningful test case?

My gut feeling is that stealing a Python reference like this is much more likely to backfire than be helpful. When dropping down to the bare-metal level of `PyObject *`, it's basically a given that the author of the code with the raw pointers needs to think 3+ times about all nuances of lifetime management. Keeping the mechanics here as simple as possible makes that easier.

Reply:
Actually, a slight modification to the current one of the current test cases would definitely cover it (see below).